James Longdev

jamesbeedy opened the pull request:
Please reference the [Slurm Charms Change Summary Document](https://docs.google.com/document/d/1bP95EXfM-1qp81IF1CA4igLVZIKnsjWdHdN6PZpEh74/edit) for an in-depth explanation of these changes.

Note: Unit tests have been mangled and are not currently passing, although the charm is working as expected. Need to spend a few minutes going through and fix the broken unit tests.

1) remove slurm-ops-manager
2) remove unused code
3) emit relation data to the charm via events
4) remove dependency on slurmctld
5) general code cleanup
6) consolidate yaml files into charmcraft.yaml
7) add type tests
8) rename interfaces


Related:
* [slurmd](https://github.com/charmed-hpc/slurmd-operator/pull/34)
* [slurmctld](https://github.com/charmed-hpc/slurmctld-operator/pull/26)
* fixes: #5

NucciTheBoss commented on src/slurmdbd_ops.py:
```diff
@@ -0,0 +1,448 @@
+        return slurmdbd_vers.version.number
+
+
+class SlurmdbdOpsManager:
```
So one thing I'm wondering here is if we should condense the API of this class here. Since the operations management is now scoped per charm, it might make sense to scope an operation manager to each service we have within the charm. For example, we could have an operations manager for both `munge` and `slurmdbd`.

Then the method signatures could shorter like the following:

```python
class SlurmdbdManager:
    def install(...): ...

    def remove(...): ...
    
    def update(...): ...
```

What do you think?

NucciTheBoss commented on src/slurmdbd_ops.py:
```diff
@@ -0,0 +1,448 @@
+        return Slurmdbd().version()
+
+
+CHARM_MAINTAINED_PARAMETERS = {
+    "DbdPort": "6819",
+    "AuthType": "auth/munge",
+    "AuthInfo": '"socket=/var/run/munge/munge.socket.2"',
+    "SlurmUser": "slurm",
+    "PluginDir": "/usr/lib/x86_64-linux-gnu/slurm-wlm",
+    "PidFile": "/var/run/slurmdbd.pid",
+    "LogFile": "/var/log/slurm/slurmdbd.log",
+    "StorageType": "accounting_storage/mysql",
+}
```
Should this possibly be in a `constants.py` module? This is a common design pattern that I have seen within other charms.

NucciTheBoss commented on .github/workflows/ci.yaml:
```diff
@@ -51,6 +51,17 @@ jobs:
       - name: Run tests
         run: tox -e unit
 
+  type-check:
+    name: Mypy tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Install dependencies
+        run: python3 -m pip install tox
+      - name: Run tests
+        run: tox -e type
```
Are we sure we want to block CI on type checking? 

To be honest, I've always been a bit conflicted about strict type checking in Python since it is a feature that is bolted onto the language rather than being directly built in like Rust or Go. I think it can be helpful for linting, but I don't think it should be a hard requirement. Reason saying is that it's easy to become lazy and add `#type: ignore` everywhere where we don't want to spend typing. It's also not necessarily enforced at runtime either. I do think type hinting as documentation is good for expected types.

Let me know you thoughts.

jamesbeedy replied:
I do like the type-checking. Omnivector goes to great lengths in other code projects to make sure the code is typed, and I have grown to like reading the code with type hints. I wouldn't say I like ignoring the type checks inline, but I was stuck and wanted to move on, so I ignored cases I couldn't solve. I should do a better job of notating my reasoning for neglecting.

If you are opposed to it, possibly we can have a conversation about it in our next community meting?

NucciTheBoss replied:
I'm not opposed to it's inclusion, in fact, I think it's good for development because it commits us to accurately documenting methods and what types they accept v.s. don't. Biggest pain is when you have a super nested method that fails because it received a type that you didn't expect it too, or you didn't realize that your object was casted into a different type by another method.

I'm fine with adding it as long as we commit to documenting why `# type: ignore` is used in certain places. Doesn't need to be super extensive, and can be a short note in the commit message. What do you think of that?

jamesbeedy replied:
Love it! Ill add some documentation to my ignores. To wrap up this pr.

NucciTheBoss reviewed (changes requested):
Just a few comments. We'll need to fix unit tests before we merge. Seems to just be some mismatch between expected status messages as they've changed with the updated methods.

jamesbeedy commented:
Thanks, @NucciTheBoss! I've addressed your comments and suggestions and resolved all but the type-check conversation. I'll move these changes forward into the slurmctld-operator next.

Thanks!

NucciTheBoss commented on src/slurmdbd_ops.py:
```diff
@@ -0,0 +1,447 @@
+#!/usr/bin/env python3
```
[nit]: Since this is a imported module, it doesn't need to have a hashbag at the top of the file

```suggestion
```

NucciTheBoss commented on src/slurmdbd_ops.py:
```diff
@@ -0,0 +1,447 @@
+
+logger = logging.getLogger()
+
+SLURMDBD_DEFAULTS = Path("/etc/default/slurmdbd")
```
Prime candidate for being moved to constants :smile:

NucciTheBoss commented on src/slurmdbd_ops.py:
```diff
@@ -0,0 +1,447 @@
+        self.message = message
+
+
+UBUNTU_HPC_PPA_KEY: str = """
+-----BEGIN PGP PUBLIC KEY BLOCK-----
+Comment: Hostname:
+Version: Hockeypuck 2.1.1-10-gec3b0e7
+
+xsFNBGTuZb8BEACtJ1CnZe6/hv84DceHv+a54y3Pqq0gqED0xhTKnbj/E2ByJpmT
+NlDNkpeITwPAAN1e3824Me76Qn31RkogTMoPJ2o2XfG253RXd67MPxYhfKTJcnM3
+CEkmeI4u2Lynh3O6RQ08nAFS2AGTeFVFH2GPNWrfOsGZW03Jas85TZ0k7LXVHiBs
+W6qonbsFJhshvwC3SryG4XYT+z/+35x5fus4rPtMrrEOD65hij7EtQNaE8owuAju
+Kcd0m2b+crMXNcllWFWmYMV0VjksQvYD7jwGrWeKs+EeHgU8ZuqaIP4pYHvoQjag
+umqnH9Qsaq5NAXiuAIAGDIIV4RdAfQIR4opGaVgIFJdvoSwYe3oh2JlrLPBlyxyY
+dayDifd3X8jxq6/oAuyH1h5K/QLs46jLSR8fUbG98SCHlRmvozTuWGk+e07ALtGe
+sGv78ToHKwoM2buXaTTHMwYwu7Rx8LZ4bZPHdersN1VW/m9yn1n5hMzwbFKy2s6/
+D4Q2ZBsqlN+5aW2q0IUmO+m0GhcdaDv8U7RVto1cWWPr50HhiCi7Yvei1qZiD9jq
+57oYZVqTUNCTPxi6NeTOdEc+YqNynWNArx4PHh38LT0bqKtlZCGHNfoAJLPVYhbB
+b2AHj9edYtHU9AAFSIy+HstET6P0UDxy02IeyE2yxoUBqdlXyv6FL44E+wARAQAB
+zRxMYXVuY2hwYWQgUFBBIGZvciBVYnVudHUgSFBDwsGOBBMBCgA4FiEErocSHcPk
+oLD4H/Aj9tDF1ca+s3sFAmTuZb8CGwMFCwkIBwIGFQoJCAsCBBYCAwECHgECF4AA
+CgkQ9tDF1ca+s3sz3w//RNawsgydrutcbKf0yphDhzWS53wgfrs2KF1KgB0u/H+u
+6Kn2C6jrVM0vuY4NKpbEPCduOj21pTCepL6PoCLv++tICOLVok5wY7Zn3WQFq0js
+Iy1wO5t3kA1cTD/05v/qQVBGZ2j4DsJo33iMcQS5AjHvSr0nu7XSvDDEE3cQE55D
+87vL7lgGjuTOikPh5FpCoS1gpemBfwm2Lbm4P8vGOA4/witRjGgfC1fv1idUnZLM
+TbGrDlhVie8pX2kgB6yTYbJ3P3kpC1ZPpXSRWO/cQ8xoYpLBTXOOtqwZZUnxyzHh
+gM+hv42vPTOnCo+apD97/VArsp59pDqEVoAtMTk72fdBqR+BB77g2hBkKESgQIEq
+EiE1/TOISioMkE0AuUdaJ2ebyQXugSHHuBaqbEC47v8t5DVN5Qr9OriuzCuSDNFn
+6SBHpahN9ZNi9w0A/Yh1+lFfpkVw2t04Q2LNuupqOpW+h3/62AeUqjUIAIrmfeML
+IDRE2VdquYdIXKuhNvfpJYGdyvx/wAbiAeBWg0uPSepwTfTG59VPQmj0FtalkMnN
+ya2212K5q68O5eXOfCnGeMvqIXxqzpdukxSZnLkgk40uFJnJVESd/CxHquqHPUDE
+fy6i2AnB3kUI27D4HY2YSlXLSRbjiSxTfVwNCzDsIh7Czefsm6ITK2+cVWs0hNQ=
+=cs1s
+-----END PGP PUBLIC KEY BLOCK-----
+"""
```
Prime candidate for being moved to constants :smile:

NucciTheBoss commented on src/slurmdbd_ops.py:
```diff
@@ -0,0 +1,447 @@
+        slurmdbd_conf.chmod(0o600)
+
+        try:
+            subprocess.call(["chown", self._user_group, f"{slurmdbd_conf}"])
```
Any particular reason why you're not using [`os.chown(...)`](https://docs.python.org/3/library/os.html#os.chown) here?

You can use the [`pwd`](https://docs.python.org/3/library/pwd.html) and [`grp`](https://docs.python.org/3/library/grp.html) modules to retrieve the uid and gid of the slurm user.

NucciTheBoss commented on src/slurmdbd_ops.py:
```diff
@@ -0,0 +1,447 @@
+        except subprocess.CalledProcessError as e:
+            raise (e)
+
+    def get_munge_key_from_fs(self) -> str:
+        """Return the munge_key from the munge.key on the filesystem."""
+        munge_key = ""
+        if self._munge_key_path.exists():
+            munge_key = b64encode(self._munge_key_path.read_bytes()).decode()
+        return munge_key
```
Should we throw an error if there's no munge key? Or is it okay for this method to return an empty string?

jamesbeedy replied:
removed this function as it is unused

NucciTheBoss reviewed (changes requested):
Just a few more suggested changes. After this I think we're good to merge. I'll just want to check against the other operators since the integration tests are relying upon incompatible charms in Charmhub currently

jamesbeedy commented:
It still needs tests fixed and commenting for type `ignore` comments.

NucciTheBoss commented:
> It still needs tests fixed and commenting for type `ignore` comments.

Okay, this is looking real good now! For the unit tests, seems like we just need to mock the outputs of `getpwname` and such. Not too difficult! For a future enhancement, we could consider looking at `pytest-mocker` for mocking objects. Has a pretty nice API, but out of scope for this PR :smile:

NucciTheBoss commented on src/slurmdbd_ops.py:
```diff
@@ -0,0 +1,395 @@
+        """Set the initial attribute values."""
+        self._slurm_component = "slurmdbd"
+
+        self._slurm_state_dir = Path("/var/spool/slurmdbd")
+        self._jwt_rsa_key_file = self._slurm_state_dir / "jwt_hs256.key"
+        self._munge_key_path = Path("/etc/munge/munge.key")
+        self._slurmdbd_conf_dir = Path("/etc/slurm")
+        self._slurmdbd_conf_file = self._slurmdbd_conf_dir / "slurmdbd.conf"
+        self._slurmdbd_log_dir = Path("/var/log/slurm")
+        self._slurmdbd_log_file = self._slurmdbd_log_dir / "slurmdbd.log"
```
I've noticed that when searching for usages of the attributes, I'm seeing that many of them are only ever read once or twice. I also see that some attributes - for example `_jwt_rsa_key_file` - are then assigned to a local variable within the method that they are called.

I think it would be simpler to create the `Path(...)` objects locally, on-the-fly within the methods that they are used. Since we're opinionated about where these files are going to be located, I don't think we need to make these values object attributes that can be redefined accidentally. Having a bunch of protected attributes that are only accessed once or twice adds additional code complexity where I think we could get away with hardcoded values.

NucciTheBoss commented on src/slurmdbd_ops.py:
```diff
@@ -0,0 +1,395 @@
+        self._slurmdbd_conf_file = self._slurmdbd_conf_dir / "slurmdbd.conf"
+        self._slurmdbd_log_dir = Path("/var/log/slurm")
+        self._slurmdbd_log_file = self._slurmdbd_log_dir / "slurmdbd.log"
+        self._slurm_user = "slurm"
+        self._slurm_group = "slurm"
+        self._user_group = f"{self._slurm_user}:{self._slurm_group}"
+        self._slurm_user_uid = getpwnam(self._slurm_user).pw_uid
+        self._slurm_group_gid = getgrnam(self._slurm_group).gr_gid
```
Once we fix this, it will be easier to complete the unit tests since we don't need to patch `getpwnam(...)` for every test :smile:

NucciTheBoss reviewed (changes requested):
Couple comments on the ops manager. I noticed some bugs in the `__init__` method that will cause the slurmdbd operator to fail to deploy :sweat_smile:

NucciTheBoss commented on .gitignore:
```diff
@@ -8,3 +8,5 @@ __pycache__/
 .vscode/
 version
+.ruff_cache/
+.mypy_cache/
```
[nit]: we don't need the .mypy_cache in .gitignore anymore since we're using pyright instead :smile: 

```suggestion
```

NucciTheBoss reviewed (approved):
I'm happy with how this pull request has shaped up! I'm ready to merge it :ship:  

Only one small nit, but I can commit that to the batch. Just removing `.mypy_cache/` from _.gitignore_ since we now using pyright instead.

We can further refine the charm in future pull requests as we stabilize the new API. I'd like to see how we can use a dedicated charm library to reduce code duplication for managing Slurm. Thank you for this work @jamesbeedy :rocket:
